### PR TITLE
10301 Fix support for `contextvar.reset`

### PR DESCRIPTION
**Pull request by richvdh**
## Scope and purpose

We need to use the same copy of the context for each iteration of the generator function, to ensure that `.reset` works as expected.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10301
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: richvdh
Reviewer: <github_username>, <github_username_if_more_reviewers>
Fixes: ticket:10301

We need to use the same copy of the context for each iteration of the generator function,
to ensure that `.reset` works as expected.
```

### CODE REVIEW

**Comment by itamarst** [comment]
I encountered the same problem (it impacts Eliot logging—https://eliot.readthedocs.io/en/stable/—which relies on contextvars). My reproducer:

```python
from twisted.internet.defer import inlineCallbacks
from twisted.internet.task import deferLater
from twisted.internet import reactor
from contextvars import ContextVar
from contextlib import contextmanager

MY_VAR = ContextVar("example")

@contextmanager
def set_var(name):
    print("setting!", name)
    token = MY_VAR.set(1)
    try:
        yield
    except Exception as e:
        print("resetting, exception!", name, e)
        MY_VAR.reset(token)
        raise
    else:
        print("resetting, success!", name)
        MY_VAR.reset(token)
        print("finished reseting", name)

@inlineCallbacks
def parent():
    with set_var("parent"):
        print("parent yielding")
        x = yield child()
        print("parent done yielding")
        return x

@inlineCallbacks
def child():
    with set_var("child"):
        print("child yielding")
        x = yield deferLater(reactor, 0.01, lambda: 123)
        print("child done yielding")
        return x

parent().addBoth(lambda r: print("Result", r) or reactor.stop())
reactor.run()
```

This PR fixes the problem, and conceptually makes sense to me as well.

File: src/twisted/test/test_defer.py [hunk 1]
```diff
@@ -3364,10 +3364,26 @@ def test_withInlineCallbacks(self) -> None:
 
         @defer.inlineCallbacks
         def yieldingDeferred() -> Generator[Deferred[Any], Any, None]:
+            # first try setting the var
```
Comment (exarkun) [inline src/twisted/test/test_defer.py]: This test is already excessively complex.  Can you please move testing for this fix into a new test method dedicated to this one specific behavior?

**Review by exarkun — CHANGES_REQUESTED** [review]
Thanks!  The implementation change looks correct to me.  I left a comment about test factoring inline.  Once that is addressed I would be happy to see this merged.

File: src/twisted/test/test_defer.py (hunk 1 above)
Comment (richvdh) [reply src/twisted/test/test_defer.py]: @exarkun thanks for the review! I've split this test up as you suggest in 83c9a66.

Please could you take another look?

**Review by exarkun — APPROVED** [review]
Thanks!

**Pull request merged**